Automattic/woocommerce-payments
# Fix E2E shopper tests around 3DS and UPE settings

This PR should fix the E2E shopper tests on pull requests.

#### Changes proposed in this Pull Request
There were two particular failures that were occurring in previous runs:


```
Saved cards  › when using a 3DS card added on checkout › should save the card

    : Timeout - Async callback was not invoked within the 100000 ms timeout specified by jest.setTimeout.Timeout - Async callback was not invoked within the 100000 ms timeout specified by jest.setTimeout.Error:
```

```
Saved cards  › when using a basic card added on checkout › should delete the card

    No node found for selector: #username
```

```
Enabled UPE with deferred intent creation › Enabled UPE with deferred intent creation › should successfully place order with Giropay

    TypeError: text is not iterable
```

The solution's I'm applying in this PR for the 3DS issue is to address the fact that the additional `'iframe[name="acsFrame"]'` is no longer present with the retirement of 3DS1. I hadn't seen any particular announcement around this change from Stripe, but it is not unexpected as our current conditional for looking for that iframe states it was a 3DS1 specific feature. I haven't been able to determine when / how this was introduced and could have been just a change from Stripe, but would be nice to narrow this down.

The solution to the missing `#username` field on the saved cards tests was actually caused by a previous tests issues. In the `shopper-checkout-purchase-with-upe-methods.spec` the `merchantWCP.disablePaymentMethod` function was not completing, causing the merchant to not be logged out, causing the next test to fail when trying to log in. This was due to the use of Xpath when the function was expecting a selector. There was already some prior art in `merchantWCP.enablePaymentMethod` which I applied here as well. This was introduced in https://github.com/Automattic/woocommerce-payments/pull/7954 when converting selectors into Xpath.

The solution to the iterable issue was to add a `postcode` to the newly added `upe-customers` data. This was introduced in https://github.com/Automattic/woocommerce-payments/pull/8082 but was likely unnoticed as the E2E tests were failing for a while by this point.

One thing I considered was whether we should look to make those tests more resilient by first confirming we're logged out. I leaned away from that just because it seemed that it may cover up issues in other tests not getting "cleaned up" properly, but wanted to open that up for some discussion (and perhaps open an issue to handle in a seperate PR?).

#### Testing instructions

- Run tests locally to confirm all tests in the shopper branch are passing: `E2E_GROUP='wcpay' E2E_BRANCH='shopper' npm run test:e2e`
- Confirm that all GH checks are passing for `[WC - latest | wcpay - shopper]`

-------------------


**Post merge**
